src/coordsim/simulation/simulatorparams.py

```diff
@@ -106,13 +126,33 @@ def update_single_inter_arr_mean(self, new_mean):
     def update_single_predicted_inter_arr_mean(self, new_mean):
         self.predicted_inter_arr_mean = {node_id: new_mean for node_id in self.network.nodes}
 
-    # string representation for logging
-    def __str__(self):
-        params_str = "Simulator parameters: \n"
-        params_str += "inter_arr_mean: {}\n".format(self.inter_arr_mean)
-        params_str += f"deterministic_arrival: {self.deterministic_arrival}\n"
-        params_str += "flow_dr_mean: {}\n".format(self.flow_dr_mean)
-        params_str += "flow_dr_stdv: {}\n".format(self.flow_dr_stdev)
-        params_str += "flow_size_shape: {}\n".format(self.flow_size_shape)
-        params_str += f"deterministic_size: {self.deterministic_size}\n"
-        return params_str
+    def generate_flow_lists(self):
+        """Generate and set dicts of lists of flow arrival, size, dr for the run duration"""
+        if self.deterministic_arrival or self.use_states or self.use_trace:
+            # TODO: implement for deterministic traffic and MMPP?
+            raise NotImplementedError()
+
+        # init empty list for each ingress node: node_id --> []
+        flow_arrival = {ing[0]: [] for ing in self.ing_nodes}
+        # TODO: init flow size & dr
+        # generate flow inter-arrival times for each ingress
+        for ing in flow_arrival.keys():
+            # generate flows for time frame of num_steps
+            while sum(flow_arrival[ing]) < self.run_duration:
+                inter_arr_time = random.expovariate(lambd=1.0/self.inter_arr_mean[ing])
+                flow_arrival[ing].append(inter_arr_time)
+                # TODO: generate flow size and dr
+
+        # set index and generated dicts of lists as attributes
+        self.flow_list_idx = 0
+        # TODO: return flow size and dr
+        self.flow_arrival_list = flow_arrival
+
+    def get_next_flow_data(self, ing):
+        # TODO: extend to get all flow data (incl size and dr)
+        """Return next flow data for given ingress from list of generated arrival times."""
+        assert self.flow_list_idx < len(self.flow_arrival_list[ing])
+        inter_arrival_time = self.flow_arrival_list[ing][self.flow_list_idx]
+        # important: increment index!
+        self.flow_list_idx += 1
```

Comment:
This won't work with mutli ingress nodes since it is an object level variable and all simpy generators will access the same memory location. This is why I set the index locally in the generator itself. My problem with it was that the code block resetting the index was not executing all the time and causing the index variable to continue over next run-duration, only for some runs and not all.

Reply:
Good point, we'll need different indices for each ingress node. I'd just make the `self.flow_list_idx` a dict instead: ingress_id --> index